In [3]:
import os
import cv2
import numpy as np
from sklearn.model_selection import train_test_split

base_path = "/kaggle/input/datasets/xiaose/cityscapes/Cityspaces/images/val"

cities = os.listdir(base_path)

X = []
y = []

for label, city in enumerate(cities):
    folder = os.path.join(base_path, city)

    for file in os.listdir(folder):
        if file.endswith(".png"):
            img = cv2.imread(os.path.join(folder, file))
            img = cv2.resize(img, (64,64))

            X.append(img)
            y.append(label)

X = np.array(X) / 255.0
y = np.array(y)

# reduce size for speed
X = X[:400]
y = y[:400]

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

LeNet

In [8]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, AveragePooling2D, Flatten, Dense, Input

model = Sequential([
    Input(shape=(64,64,3)),   #  fix warning also

    Conv2D(6, (5,5), activation='relu'),
    AveragePooling2D(pool_size=(2,2)),

    Conv2D(16, (5,5), activation='relu'),
    AveragePooling2D(pool_size=(2,2)),

    Flatten(),
    Dense(120, activation='relu'),
    Dense(84, activation='relu'),
    Dense(3, activation='softmax')
])

model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

model.fit(X_train, y_train, epochs=15, verbose=1)

print("LeNet Accuracy:", model.evaluate(X_test, y_test)[1])

Epoch 1/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 2s 53ms/step - accuracy: 0.4431 - loss: 1.0289
Epoch 2/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 53ms/step - accuracy: 0.4831 - loss: 0.9748
Epoch 3/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 53ms/step - accuracy: 0.5202 - loss: 0.9166
Epoch 4/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy: 0.6745 - loss: 0.7966
Epoch 5/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 52ms/step - accuracy: 0.7520 - loss: 0.6548
Epoch 6/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 55ms/step - accuracy: 0.7395 - loss: 0.6219
Epoch 7/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy: 0.7892 - loss: 0.5059
Epoch 8/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 52ms/step - accuracy: 0.7692 - loss: 0.4997
Epoch 9/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy: 0.8202 - loss: 0.4114
Epoch 10/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 50ms/step - accuracy: 0.8560 - loss: 0.3187
Epoch 11/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy: 0.8593 - loss: 0.2976
Epoch 12/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 1s 51ms/step - accuracy:

**Alexnet**

In [24]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, Input
from tensorflow.keras.utils import to_categorical

# 🔹 One-hot encoding (same as your MNIST code)
y_train_cat = to_categorical(y_train, 3)
y_test_cat = to_categorical(y_test, 3)

# 🔹 Build AlexNet
model = Sequential()

model.add(Input(shape=(64,64,3)))

# Conv Layer 1
model.add(Conv2D(96, (5,5), strides=1, activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2)))

# Conv Layer 2
model.add(Conv2D(256, (3,3), activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2)))

# Conv Layers 3,4,5
model.add(Conv2D(384, (3,3), activation='relu', padding='same'))
model.add(Conv2D(384, (3,3), activation='relu', padding='same'))
model.add(Conv2D(256, (3,3), activation='relu', padding='same'))
model.add(MaxPooling2D(pool_size=(2,2)))

# Fully connected
model.add(Flatten())

model.add(Dense(512, activation='relu'))
model.add(Dropout(0.5))

model.add(Dense(256, activation='relu'))
model.add(Dropout(0.5))

model.add(Dense(3, activation='softmax'))

# 🔹 Compile
model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

# 🔹 Train
model.fit(
    X_train, y_train_cat,
    epochs=15,
    batch_size=32,
    validation_data=(X_test, y_test_cat),
    verbose=1
)

# 🔹 Evaluate
loss, acc = model.evaluate(X_test, y_test_cat)

print("AlexNet Accuracy:", acc)

Epoch 1/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 27s 2s/step - accuracy: 0.4462 - loss: 1.1357 - val_accuracy: 0.3625 - val_loss: 1.0129
Epoch 2/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 24s 2s/step - accuracy: 0.4291 - loss: 1.0240 - val_accuracy: 0.3625 - val_loss: 0.9992
Epoch 3/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 24s 2s/step - accuracy: 0.4380 - loss: 0.9956 - val_accuracy: 0.3625 - val_loss: 0.9594
Epoch 4/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 41s 2s/step - accuracy: 0.5142 - loss: 0.8651 - val_accuracy: 0.5125 - val_loss: 1.0502
Epoch 5/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 25s 2s/step - accuracy: 0.4644 - loss: 1.0266 - val_accuracy: 0.5250 - val_loss: 1.0168
Epoch 6/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 25s 3s/step - accuracy: 0.4069 - loss: 1.0520 - val_accuracy: 0.6125 - val_loss: 0.9306
Epoch 7/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 24s 2s/step - accuracy: 0.6234 - loss: 0.9346 - val_accuracy: 0.7000 - val_loss: 0.8994
Epoch 8/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 25s 3s/step - accuracy: 0.6470 - loss: 0.8791 - val_accuracy: 0.6625 - val_loss:

In [25]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, Input
from tensorflow.keras.utils import to_categorical

# one-hot
y_train_cat = to_categorical(y_train, 3)
y_test_cat = to_categorical(y_test, 3)

model = Sequential([
    Input(shape=(64,64,3)),

    Conv2D(32, (3,3), activation='relu', padding='same'),
    Conv2D(32, (3,3), activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Conv2D(64, (3,3), activation='relu', padding='same'),
    Conv2D(64, (3,3), activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Conv2D(128, (3,3), activation='relu', padding='same'),
    Conv2D(128, (3,3), activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Flatten(),
    Dense(128, activation='relu'),
    Dropout(0.4),

    Dense(3, activation='softmax')
])

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

model.fit(X_train, y_train_cat, epochs=15, batch_size=32, validation_data=(X_test, y_test_cat))

print("VGG Accuracy:", model.evaluate(X_test, y_test_cat)[1])

Epoch 1/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 7s 495ms/step - accuracy: 0.4373 - loss: 1.0821 - val_accuracy: 0.5500 - val_loss: 0.9584
Epoch 2/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 462ms/step - accuracy: 0.4890 - loss: 0.9995 - val_accuracy: 0.3625 - val_loss: 0.9787
Epoch 3/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 459ms/step - accuracy: 0.4802 - loss: 0.9351 - val_accuracy: 0.4625 - val_loss: 0.9154
Epoch 4/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 460ms/step - accuracy: 0.5957 - loss: 0.8534 - val_accuracy: 0.6250 - val_loss: 0.9221
Epoch 5/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 455ms/step - accuracy: 0.6469 - loss: 0.8763 - val_accuracy: 0.6250 - val_loss: 0.9683
Epoch 6/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 467ms/step - accuracy: 0.6872 - loss: 0.8289 - val_accuracy: 0.6250 - val_loss: 0.8685
Epoch 7/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 455ms/step - accuracy: 0.6196 - loss: 0.8349 - val_accuracy: 0.6500 - val_loss: 0.7981
Epoch 8/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 5s 459ms/step - accuracy: 0.7395 - loss: 0.6377 - val_accuracy: 0.

ZFNet

In [29]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, Input
from tensorflow.keras.utils import to_categorical

# one-hot
y_train_cat = to_categorical(y_train, 3)
y_test_cat = to_categorical(y_test, 3)

model = Sequential([
    Input(shape=(64,64,3)),

    Conv2D(96, (7,7), strides=2, activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Conv2D(256, (5,5), activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Conv2D(384, (3,3), activation='relu', padding='same'),
    Conv2D(384, (3,3), activation='relu', padding='same'),
    Conv2D(256, (3,3), activation='relu', padding='same'),
    MaxPooling2D((2,2)),

    Flatten(),

    Dense(512, activation='relu'),
    Dropout(0.5),

    Dense(256, activation='relu'),
    Dropout(0.5),

    Dense(3, activation='softmax')
])

model.compile(
    optimizer='adam',
    loss='categorical_crossentropy',
    metrics=['accuracy']
)

model.fit(
    X_train, y_train_cat,
    epochs=15,
    batch_size=32,
    validation_data=(X_test, y_test_cat)
)

print("ZF-Net Accuracy:", model.evaluate(X_test, y_test_cat)[1])

Epoch 1/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 11s 898ms/step - accuracy: 0.3979 - loss: 1.1361 - val_accuracy: 0.5625 - val_loss: 1.0238
Epoch 2/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 859ms/step - accuracy: 0.4411 - loss: 1.0682 - val_accuracy: 0.5125 - val_loss: 1.0358
Epoch 3/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 860ms/step - accuracy: 0.4110 - loss: 1.0638 - val_accuracy: 0.5125 - val_loss: 0.9871
Epoch 4/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 864ms/step - accuracy: 0.4221 - loss: 1.0371 - val_accuracy: 0.3625 - val_loss: 0.9882
Epoch 5/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 863ms/step - accuracy: 0.3981 - loss: 0.9915 - val_accuracy: 0.3625 - val_loss: 1.0037
Epoch 6/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 865ms/step - accuracy: 0.4388 - loss: 1.0449 - val_accuracy: 0.3625 - val_loss: 1.0072
Epoch 7/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 863ms/step - accuracy: 0.4427 - loss: 1.0229 - val_accuracy: 0.3625 - val_loss: 0.9870
Epoch 8/15
10/10 ━━━━━━━━━━━━━━━━━━━━ 9s 863ms/step - accuracy: 0.4384 - loss: 1.0119 - val_accuracy: 0

FINAL OBSERVATIONS

* LeNet achieved the highest accuracy (~83.7%) due to its simple architecture, which is well-suited for small datasets.

* AlexNet showed slightly lower accuracy (~80%) but performed well due to deeper layers and better feature extraction.

* VGG model achieved moderate accuracy (~72.5%) as deeper architectures require more data and are harder to train from scratch.

* ZF-Net performed poorly (~36%) due to excessive complexity and reduction in feature map size, leading to ineffective learning on a small dataset.